File: src/main/java/net/serebryansky/common/service/impl/RoutingServiceImpl.java
package net.serebryansky.common.service.impl;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.serebryansky.common.service.RoutingService;
import net.serebryansky.place.model.Place;
import net.serebryansky.routing.model.DurationMatrixResponse;
import net.serebryansky.routing.model.Route;
import net.serebryansky.routing.model.RouteSearchRequest;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.List;

import static net.serebryansky.common.util.LoggingUtil.logOnNext;

@Slf4j
@RequiredArgsConstructor
public class RoutingServiceImpl implements RoutingService {
    private final WebClient routingClient;

    @Override
    public Mono<DurationMatrixResponse> getDurationMatrix(List<Place> p) {
        return Mono.just(p)
                .doOnEach(logOnNext(p1 -> log.info("Get duration matrix for {}", p1)))
                .then(Mono.deferContextual(contextView -> {
                    String requestId = contextView.get("CONTEXT_KEY");
                    String invoker = contextView.get("INVOKER");
                    return routingClient.post()
                            .uri("/durations/matrix")
                            .header("X-Request-ID", requestId)
                            .header("X-Source", invoker)
                            .bodyValue(p)
                            .retrieve()
                            .bodyToMono(DurationMatrixResponse.class);
                }));
    }

    @Override
    public Mono<Route> getRoute(List<Place> placePair) {
        // log.info("Get route between {} and {}", placePair.get(0), placePair.get(1));
        return Mono.just(placePair)
                .doOnEach(logOnNext(p -> log.info("Get route between {} and {}", p.get(0), p.get(1))))
                .then(Mono.deferContextual(contextView -> {
                    String requestId = contextView.get("CONTEXT_KEY");
                    String invoker = contextView.get("INVOKER");
                    return routingClient
                            .post()
                            .uri("/routes/search")
                            .header("X-Request-ID", requestId)
                            .header("X-Source", invoker)
                            .bodyValue(new RouteSearchRequest(placePair.get(0), placePair.get(1)))
                            .retrieve()
                            .bodyToMono(Route.class);
                }));
    }
}

File: src/main/kotlin/net/serebryansky/common/service/impl/RoutingServiceImpl.kt
package net.serebryansky.common.service.impl

import kotlinx.coroutines.reactive.awaitFirst
import kotlinx.coroutines.reactor.mono
import mu.KotlinLogging
import net.serebryansky.common.requestKey
import net.serebryansky.common.service.RoutingService
import net.serebryansky.place.model.Place
import net.serebryansky.routing.model.DurationMatrixResponse
import net.serebryansky.routing.model.Route
import net.serebryansky.routing.model.RouteSearchRequest
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import reactor.core.publisher.Mono

class RoutingServiceImpl(
    private val routingClient: WebClient,
    private val applicationName: String,
) : RoutingService {

    private val log = KotlinLogging.logger {}

    override suspend fun getDurationMatrix(places: List<Place>): DurationMatrixResponse {
        log.info { "Get duration matrix for $places" }
        val requestId = Mono.deferContextual { mono { it.get<String>(requestKey) } }.awaitFirst()
        return routingClient
            .post()
            .uri("/durations/matrix")
            .header("X-Request-ID", requestId)
            .header("X-Source", applicationName)
            .bodyValue(places)
            .retrieve()
            .awaitBody()
    }

    override suspend fun getRoute(places: List<Place>): Route {
        log.info { "Get route between ${places[0]} and ${places[1]}" }
        val requestId = Mono.deferContextual { mono { it.get<String>(requestKey) } }.awaitFirst()
        return routingClient
            .post()
            .uri("/routes/search")
            .header("X-Request-ID", requestId)
            .header("X-Source", applicationName)
            .bodyValue(RouteSearchRequest(places[0], places[1]))
            .retrieve()
            .awaitBody()
    }
}
